Issue 2987: Update Pravega base Docker images to Alpine

adrianmo opened the pull request:
**Change log description**  
- Replace base Docker images for Pravega and BookKeeper from `openjdk:8-jre` to `openjdk:8-jre:alpine` to use Alpine-based images. 
- Use `apk` instead of `apt-get` to install system packages.
- Reduce Pravega Docker image size from 758MB to 241MB (-68%)
- Reduce BookKeeper Docker image size from 777MB to 206MB (-73%)
- Add bash flag `set -eo pipefail` to Docker entrypoints to immediately fail if a command fails. 
- Append `|| :` to entrypoint commands that are allowed to fail.

**Purpose of the change**  
- Fixes #2987
- Fixes #3008

**What the code does**  
Changes the Docker base image in:
- `docker/pravega/Dockerfile`
- `docker/bookkeeper/Dockerfile`

**How to verify it**  
Build, deploy Docker images, and run all tests based on the new images.

Signed-off-by: Adrian Moreno <adrian@morenomartinez.com>

adrianmo commented:
Need to run all and pass all tests using the images before this merging this PR.

shrids commented on docker/bookkeeper/Dockerfile:
```diff
@@ -8,11 +8,13 @@
 
-RUN apt-get update \
-&&  apt-get install -y wget unzip \
+RUN apk add --update \
```
Do not see the option `--update` inside `apk add -h`. I logged into the container openjdk:8-jre-alpine and ran the apk command.  `apk-get update` is a valid command.

adrianmo replied:
It's a shorthand for doing first `apk update` followed by `apk add [package]`. It works fine and many other projects are using it, but as you say, it's not documented...

adrianmo commented:
@shrids great catch! I missed those issues because the entrypoint script did continue instead of the errors. I've added the missing dependencies and also added the `set -e` flag to the entrypoint scripts so that they exit if any of the instructions inside fail, otherwise we may fail to spot new issues again.

arvindkandhare commented on docker/pravega/Dockerfile:
```diff
@@ -7,10 +7,13 @@
+RUN apk add --update \
+    rpcbind \
+    nfs-utils \
+    python \
```
I see that `python` is added here whereas it was not present in the earlier Dockerfile. Please validate whether we need this.

EronWright replied:
@adrianmo sorry to ping an old conversation, but wasn't the real question, "does Pravega require Python"?

adrianmo replied:
@EronWright Python was added because there's a Python script that needs it. However, I've created https://github.com/pravega/pravega/issues/3251 to replace that Python script with a Bash alternative and eliminate the Python dependency.

arvindkandhare commented on docker/pravega/entrypoint.sh:
```diff
@@ -9,6 +9,8 @@
 #     http://www.apache.org/licenses/LICENSE-2.0
 #
 
+set -e
```
There are two possibilities with a dockerfile. Either keep on running in case of error or exit the startup so that container can be rescheduled somewhere else by the deploying tech. Looks like we are taking the second route. Can you comment more about the decision to go with this option?

adrianmo replied:
My reasoning is that we might miss fatal errors if we keep running in case of error. This has just happened with the Python dependency and might happen again. It's much safer to fail fast instead of ignoring the error and running in a potentially degraded mode.

adrianmo commented:
@shrids We can't use `sh` instead of `bash` in the BookKeeper image because we are using some ZooKeeper utility scripts (`zkCli.sh`, ...) that depend on `bash`.

adrianmo commented:
We are having a network issue in Jenkins when the job tries to pull the new images that causes the build to fail. Please do not merge the PR until we are able to run the system tests and a longevity test.

adrianmo commented:
After a few changes in the PR, system tests have passed. 

@fpj @arvindkandhare @shrids can you please give it a second round of reviews?

shrids commented on docker/bookkeeper/entrypoint.sh:
```diff
@@ -9,6 +9,8 @@
 #     http://www.apache.org/licenses/LICENSE-2.0
 #
 
+set -eo pipefail
```
+1

adrianmo replied:
These flags try to make the script safer:

- `-e` option causes a bash script to exit immediately when a command fails.
- `-o pipefail` option sets the exit code of a pipeline to that of the rightmost command to exit with a non-zero status. Or to zero if all commands of the pipeline exit successfully.

shrids commented on docker/pravega/Dockerfile:
```diff
@@ -7,10 +7,14 @@
+RUN apk add --update \
+    rpcbind \
+    nfs-utils \
+    libc6-compat \
```
Did not understand the reason for this change?

adrianmo replied:
Since we are changing the base image to Alpine, we need to use the `apk` package manager instead of `apt`. Package names for Alpine may differ to Debian packages. In the case of the Pravega image, we made the following changes:

- `rpcbind` Alpine package replaces the `rpcbind` Ubuntu package
- `nfs-utils` Alpine package replaces the `nfs-common` Ubuntu package
- `libc6-compat` Alpine package contains the `ld-linux-x86-64.so.2` library, which is a dependency of RocksDB. `libc6` was already installed in the previous base image.

arvindkandhare commented on docker/bookkeeper/entrypoint.sh:
```diff
@@ -61,14 +63,14 @@ echo "wait for zookeeper"
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME}
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME}/${BK_CLUSTER_NAME}
+# Silence exit codes with "|| :" as the commands can safely fail with a "Node already exists" error
```
This is dangerous as only the specific error needs to be ignored. We should continue failing in case of all the other errors.

adrianmo replied:
The problem is that `zkCli.sh` does not return specific error codes based on the error cause. Realistically, the command should not throw any other error in this scenario. The connection to the ZK server is checked right before those lines, and the only potential error is that the path already exists... In any case, this is an improvement compared to the previous script version, where all commands could fail silently.

arvindkandhare replied:
I agree that this is an improvement over previous status. Please spend some time to figure out a way to take care of this better. If there is no easy way, let's log an issue to track this.

adrianmo replied:
@arvindkandhare I'll create a new issue to track this. But I would like to have this PR merged first so that I can include proper reference to the script commands.

fpj replied:
Post the issue number here, please.

adrianmo replied:
https://github.com/pravega/pravega/issues/3039

arvindkandhare commented on docker/bookkeeper/entrypoint.sh:
```diff
@@ -61,14 +63,14 @@ echo "wait for zookeeper"
 echo "format the bookie"
-# format bookie
-BOOKIE_CONF=${BK_HOME}/conf/bk_server.conf ${BK_HOME}/bin/bookkeeper shell metaformat -nonInteractive
+# Silence exit codes with "|| :" as the command can safely fail if another instance has already formatted the bookie
```
Same as above.

fpj replied:
Post the issue number you created to track the conversation around silencing exit codes.

adrianmo replied:
https://github.com/pravega/pravega/issues/3039

fpj commented on docker/compose/docker-compose.yml:
```diff
@@ -87,6 +87,7 @@ services:
     command: segmentstore
     environment:
       WAIT_FOR: bookie1:3181,bookie2:3182,bookie3:3183,hdfs:8020
+      TIER2_STORAGE: "HDFS"
```
Out of curiosity, why didn't we need to specify tier 2 storage before, and now we do?

adrianmo replied:
My assumption is that this particular Docker compose file has not been used since the `TIER2_STORAGE` was introduced. @arvindkandhare can you confirm that?

arvindkandhare replied:
This is right. The assumption is that the default is `HDFS` but it was changed to `INMEMORY`. It is safer to explicitly mention it here.

fpj commented on docker/bookkeeper/entrypoint.sh:
```diff
@@ -61,14 +63,14 @@ echo "wait for zookeeper"
+# Silence exit codes with "|| :" as the commands can safely fail with a "Node already exists" error
+${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH} || :
+${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME} || :
+${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME}/${BK_CLUSTER_NAME} || :
```
You sure there is no way to submit three commands through the same session with the CLI, rather than doing one session per command? That should be a simple addition.

adrianmo replied:
The [ZooKeeper Create Node API](https://zookeeper.apache.org/doc/r3.4.6/api/org/apache/zookeeper/ZooKeeper.html#create(java.lang.String,%20byte[],%20java.util.List,%20org.apache.zookeeper.CreateMode)) does not allow to create nodes in paths that do not exist. To do this, we would need to wrap it in a function, but I don't think it would add any value to the current solution.

fpj replied:
That's not what I meant. We should be able to do several create operations over the same session. Each execution of zkCli will create a new session unnecessarily.
